### ISSUE TITLE
Scheduling bundles sample-accurately relative to each other

### ISSUE THREAD

**Issue by snappizz** [supercollider/supercollider#2939]
is there any way to schedule OSC messages with a sample-accurate offset between them?

it's possible in NRT mode. here is a proof of concept where i take two cosine waves of exactly 3 periods and concatenate them -- i see no discontinuity when inspecting the output file:

```supercollider
(
SynthDef(\test, { |dur = 0|
    var time;
    time = Sweep.ar;
    FreeSelf.kr(time > (dur + ControlDur.ir));
    OffsetOut.ar(0, SinOsc.ar(440, pi/2) * (time < dur));
}).writeDefFile;
)

(
Score.recordNRT(
    [
        [0, [\s_new, \test, 1000, 0, 0, \dur, 3 / 440]],
        [3 / 440, [\s_new, \test, 1001, 0, 0, \dur, 3 / 440]],
        [6 / 440, [0]]
    ],
    Platform.defaultTempDir +/+ "myscore.osc",
    "~/test.wav".standardizePath,
    options: ServerOptions().blockSize_(64),
    sampleRate: 48000,
);
)
```

but it doesn't seem feasible in real time.

```supercollider
(
SynthDef(\test, { |dur = 0|
    var time;
    time = Sweep.ar;
    FreeSelf.kr(time > (dur + ControlDur.ir));
    OffsetOut.ar(0, SinOsc.ar(440, pi/2) * 0.1 * (time < dur));
}).add;
)

(
s.makeBundle(s.latency, { Synth(\test, [\dur, 100 / 440]) });
s.makeBundle(s.latency + (100 / 440), { Synth(\test, [\dur, 100 / 440]) });
)
```

i think this is because the second call to `s.makeBundle` is slightly later than the first.

is there any workaround?

**Comment by crucialfelix** [comment]
afaict your example should work as advertised.

I did notice while writing supercollider.js that sclang assumes only delta from now (when the code is executed) and converts this to OSC time (NTP time tag array). Rather than have some way to specify an OSC time and get an exact non-floating point time and not "relative to right about now".

```markdown
 * Received OSC bundles that were converted with `fromBuffer` will have a timetag array:
 * `[secondsSince1970, fractionalSeconds]`
 *
 * That has an accuracy of 0.00000000023283 seconds or 2^32 per second or 4,294,979,169.35102864751106 per second.
 *
 * Note that the sample rate for audio is usually only 44.1kHz.
```

That is two 32 bit integers.

**Comment by telephon** [comment]
The timing should be accurate, that is, the time stamps *should* be the same for two bundles sent at the same logical time.

http://doc.sccode.org/Guides/ServerTiming.html

If this is not the case, we are either missing something or this is a bug.

**Comment by jamshark70** [comment]
> The timing should be accurate, that is, the time stamps should be the same for two bundles sent at the same logical time.

That is correct, but Nathan's example uses two different latency values -- `s.latency` and `s.latency + (100 / 440)` -- so it is true but irrelevant in this case.

I think the problem is related to this:

http://doc.sccode.org/Classes/ServerOptions.html#-useSystemClock (valid only for supernova btw)

That is, the assumption here is that it's a scheduling problem, but the real problem is that you're not getting exactly 44100 samples per second -- it's off by a small amount (`s.actualSampleRate`).

Someone was working on a SampleClock class, but I don't remember whatever happened to it.

In supernova, you can set `s.options.useSystemClock = false` (I think) and it will interpret OSC time tags in terms of samples, not system seconds. But if you do this, an hour later, 200 ms latency will no longer be 200 ms.

**Comment by jamshark70** [comment]
Oh, I should clarify, "valid only for supernova btw" means scsynth can sync only to the system clock, same as `true` for supernova. Currently, AFAIK, truly sample-accurate scheduling is impossible with a real time scsynth.

I just tried both system-clock true and false with Nathan's example:

- true, using system clock, I reproduce the problem (click between synths).

- false, using sample clock, I get a smooth sinewave.

So I can confirm, the current workaround is to use supernova with `s.options.useSystemClock = false`, and accept that timing will drift after some time.

**Comment by gogobd** [comment]
I was talking abot this with @snappizz on Gitter before this issue was created; It sounds like an easy solution to set "s.options.useSystemClock = false", however since I'm bound to Raspberry Pi I currently can't run supernova, right? It would be great to have a SampleClock because it would elegantly enable sample accurate timing in Supercollider.

**Comment by jamshark70** [comment]
FWIW I did a search yesterday for "SuperCollider SampleClock" but couldn't find it. I have a *vague* recollection that somebody forked the SC sources and worked on it, but I don't remember who. I think @muellmusik was involved in some of those discussions (maybe?)... but that was a few years ago. I'm pretty sure it never reached a point where it could be submitted as a pull request.

The trouble with a SampleClock is, it's a fairly invasive change. Clocks run "behind the scenes" -- it would be impossible to have a SampleClock without C++ support. So it's not quarkable.

My guess is:

- A true SampleClock is the best solution, but also hard to implement, so it's unlikely to happen unless somebody really, really needs it and that person also understands the internals of SC's schedulers. I wouldn't hold my breath for this.
  - The server would have to communicate precise information about samples vs time to the language. I guess that would be easy for the internal server, and *maybe* possible for a local server through the shared memory interface... maybe. Obviously it would be impossible for a remote server, so we would have to set appropriate expectations.

- If that's not practical, it would be easier to replicate supernova's `-C` flag in scsynth. You would still have the problem of drift, but that would open it up to systems where supernova isn't supported.

**Comment by snappizz** [comment]
is there some way to patch the drift problem? @gogobd's use case concerns playback of 1-hour audio files, concatenating them as accurately as possible.

**Comment by jamshark70** [comment]
The problem is that the language needs to know exactly how long the last *n* samples took, and then adjust the clock's speed.

That is, the drift is because the language assumes 1 second always == 1 second. But if you're measuring time in samples (as you must do for sample accuracy), then 44100 samples is *not* exactly 1 second. If, for instance, in a given 1-second interval, the soundcard ran slightly fast, the language needs to know exactly how much faster, so that `1.0.wait` will wait just the right amount less than a second.

In theory it's possible, but I find it difficult to imagine this is the kind of problem you can just "patch." It's not really a SuperCollider bug. It's a bad interaction between the limitations of hardware accuracy, system timing, and inter-process communication. It's not transparent.

**Comment by jamshark70** [comment]
Also, I observed yesterday that `actualSampleRate` fluctuates, so you can't predict the drift and compensate by multiplying scheduling deltas by some factor. AFAICS it needs to adjust the length of a second: you do `1.0.wait`, and logical time advances by 1.0 exactly, but the precise physical duration is determined by the server.

I think any other approach would be fudging it, and fudging seems incompatible with the idea of sample accuracy...

**Comment by gogobd** [comment]
Thanks for everyones thoughts on this issue. Would a "SampleClock" have to read the actual playback machineries time? If so I wonder if a not-so-invasive approach is conceivable - by just "handing in" this value via C++ so it can be read as a Server variable; the rest could be "quarkable"?

**Comment by telephon** [comment]
wwouldn't it be possible to query the server for how many frames were played in the previous second? A clock could interpolate and adjust, couldn't it?

**Comment by brianlheim** [comment]
I'm curious about what the use-case is here. I've always found workarounds to precise timing issues by having the server schedule its own events.

> So I can confirm, the current workaround is to use supernova with s.options.useSystemClock = false, and accept that timing will drift after some time.

@jamshark70, does this mean that `useSystemClock` is a _workaround_, or a limited-scope _solution_? i.e. is it guaranteed to work or does it just improve accuracy somewhat? I am having a hard time understanding that from this conversation.

> A true SampleClock is the best solution, but also hard to implement

What is the difference between a "true" SampleClock and the current implementation (in supernova? in scsynth?)

> The trouble with a SampleClock is, it's a fairly invasive change

Would it be implemented within the language, or within the server? I know these are basic questions but I'm trying to keep up with this conversation :)

**Comment by telephon** [comment]
the basic situation is that we have two independently running physical clocks: the system clock and the soundcard's clock. In principle, both can drift. Sclang schedules messages extremely precisely using the systen time, and the server logic makes sure the audio of the first sample hits the wire exactly at that system time. But the audio stream is varying in speed. So sclang doesn't know what the system time of the n- th sample is.

**Comment by telephon** [comment]
that is, we need some information from the soundcard directly in sclang, in order to know how long in seconds the time is when the n- th sample happens.

One unclear point is how and how often this relation has to be adjusted.

**Comment by brianlheim** [comment]
> we need some information from the soundcard directly in sclang

I am still unclear about _why_ this is absolutely necessary though. Surely having one process in charge of exact scheduling is easier than two processes trying to schedule in tandem?

**Comment by jamshark70** [comment]
> Surely having one process in charge of exact scheduling is easier than two processes trying to schedule in tandem?

The separation is implicit in the server-client model, isn't it? The alternative is to integrate the language into the DSP cycles, as SC2 did and ChucK still does, but that would make it architecturally harder to support multiclient or network-distributed applications. In theory, we could do something like that for the internal server (which I already mentioned in passing) and then say sample-accuracy works only with the internal server. (Supporting sample accuracy across a network... that's probably unrealistic.) It seems to me that it would be somewhat risky to couple the language to the internal server's DSP because you would need two versions of the language's main loop, right? One for internal and one for "normal" use... that's getting weird. If it were me, I wouldn't go that direction.

The soundcard's sample clock must be the master for sample accuracy. You couldn't have a master timing process and slave both language and server to it, because (as far as I know) slaving a soundcard to another timing source just doesn't work. We're talking about interrupts happening 50x, 100x, 200x a second (with a 256 sample hardware buffer, 44100/256 = 172.265). There's no way you can slave that and expect any accuracy. In an interprocess context, Julian's suggestion is really the only one that makes sense: server connects to audio driver, server gets timing from audio driver, language gets timing from the server.

**Comment by snappizz** [comment]
i've learned a lot of interesting things from this thread, but i think we're straying too far from the original issue. i have an idea for a (conceptually) simple solution that takes care of this particular use case of concatenating long audio files.

we introduce an OSC message `/precise`, which is of the form

    /precise [id] [time] [an osc message]

if time is negative, then the osc message is executed as if it was sent normally, and a new sample clock is created on the server with the given id. if time is nonnegative, then the server looks for an existing sample clock with that id, and then queues the OSC message, executing it when that sample clock reaches that number of samples. if the clock has already passed the scheduled time, better late than never.

so, for example, if you send the osc messages

    /precise 0 -1 /foo
    /precise 0 1000 /bar

then `/foo` will execute whenever it arrives, and `/bar` will execute exactly 1000 samples after `/foo` if it's not too late (and you can guarantee it won't be late by sending these two messages in a bundle).

the client doesn't have to stay synchronized with the sample clocks at all, and its only role is to add messages to the queue.

**Comment by telephon** [comment]
just a note: `precise` is misleading: usually, the audio hardware clock is much less precise than that system clock. So it actually less precise to schedule relative to the sample frames, and the problem to be solved is how to account for that imprecision.

I wonder why we need any server side clocks, we already have the absolute number of frames since the server started. If the client could tell the server to call a command exactly at a certain number of frames, instead of at a certain time, then this would be it.

**Comment by Sciss** [comment]
Please do not implement anything ad-hoc without thinking about the overall integration. For example, the `/precise` command is an interesting approach, but we need a way to poll the current sample clock from the client as well, if for example we implement a transport on the client, say a timeline or multi-track editor. If you can poll that clock, you can create a reasonable servo mechanism that can schedule events in the future while minimising latency.

**Comment by Sciss** [comment]
On a related note, this topic comes up roughtly every 1 1/2 years, I think I have seen already four discussions of it. Brain storming is good, but chances are that the thread will lead to no sustainable solution, if it doesn't result is some form of "specification document". I think this is what we must strive for: a specification process for solving the greater architectural issues, such as this ("sample accurate scheduling") and others ("full-fledged non-real-time mode").

It all leads back to the initiative back a few years of assembling an sc-dev group that looks for means to fund projects to realise such larger architectural changes and enhancements.

**Comment by muellmusik** [comment]
> this topic comes up roughtly every 1 1/2 years

IIRC one of the reasons it stalls is that someone said that the sample count for a device can be arbitrarily reset on some OSs. I don't know how much of an issue that would be in practice.

The problem as a whole and likely approach I think are well understood. We need a way of:
- scheduling to a sample time base
- allowing client and server access to this time base (this will only work with servers on the client machine)
- support the existing clock semantics (sched in secs), but perhaps with additional methods to allow scheduling in samples from logical now.

Am I missing anything?

**Comment by Sciss** [comment]
Sounds about right. The question is if someone feels competent enough to carry out the a prototype implementation, if it needs a group of people, etc. In order words, to organise this effort. Probably it should be done within a "topic branch", complete with specification, unit tests etc, cross-platform validation etc. That's why I'm not sure that extending it on the issue tracker is the best way.

**Comment by Sciss** [comment]
Another remark... Julian writes:

> we need some information from the soundcard directly in sclang

And Scott:

> this will only work with servers on the client machine

I would object to both these assumptions. I would very much prefer a specification for server that does not require that the sample clock functionality is duplicated on the client side. I think a servo mechanism where the client maintains a relation between _its own system clock_ and the _server's sample clock_ is much better. It makes it possible to use sample accurate scheduling on any client, not just sclang and not just clients that wire into platform specific native sound APIs, and it makes it possible to use it irrespective of the network location of the server.

----

I think the best next step would be to collate all the information pertaining to this topic from the various threads over the years in one dedicated place, perhaps with a collaborative document editor to work out a specification step by step. I don't know what the best tool is for this. Another thing that might help is that @LFSaw will be visiting the IEM in Graz soon. Till, correct me if I'm wrong, but I think it was you who suggested that the idea of a dev-group working on such projects could be revived? Perhaps with a concerted effort it could be done.

**Comment by snappizz** [comment]
i agree with @Sciss that there is some overthinking going on here, and we're getting hung up on the involvement of sclang. i still believe the best proposal in the short term is to develop a sample-accurate clock on the server only. the client can create a sample clock over OSC, schedule OSC messages on the clock, and query information about the clock's status, but technically has no obligation to stay synchronized.

we don't need to retool anything in sclang's timing model right now, we can just make this functionality available on scsynth/supernova so that advanced users with the need for sample precision can use it.

**Comment by jamshark70** [comment]
I don't have a detailed comment other than: this feels like an ad hoc workaround and not a long term sustainable solution... wherein, the problem is that new server commands are basically permanent. It'll be hard to get rid of these commands if, at a later date, someone does work on a transparent solution.

It may still be better to have some approach rather than nothing at all. I just wanted to note that syncing the language clock to the sample count would make these commands redundant.

This approach would be 100% fine to prototype in somebody's fork.

**Comment by snappizz** [comment]
i honestly don't see what's wrong with it. it may be the only solution for sample-accuracy across a network.

**Comment by Sciss** [comment]
> i honestly don't see what's wrong with it

That may be the case; and "ad-hoc" looking at it, I also don't see anything basically wrong with it. But as James said, it's in the nature of these things, that you will see all the issues once you start implementing them. It's a classical case where an algorithm might look correct by just imagining it, but when put to practice all sorts of new circumstances arise.

Anyway, I'm not against conducting an experiment, and seeing how far this gets. From then it's another step of deciding how the server commands should be called etc. (Personally, I would say they should go along with the existing ones, e.g. `[clk_new, ...]`, `[clk_get, ..]` etc.

One problem you already mentioned: numeric resolution. Why not use this as an instance to say, look we need 64bit integers. Shouldn't be too complicated to have them at the receiving side (liblo or whatever is used surely supports them). The problem is on the client side where getting 64bit ints will be quite complicated.

> but with only a master clock, you first have to query the clock, wait for a reply, then schedule the two messages relative to the reply you get, plus latency.

Scheduling multiple things at the same time, with them happening seconds apart, I think that's a very _rare_ use case. I think the standard use case is the - indeed more difficult to implement - one, where over long periods of time, you need to be able to schedule new events and keep in sync with that master clock. Obviously that means not to query the master clock "just before you want to send something", but you need an instance that provides a client clock continuously, i.e. a routine or primitive that receives timing information on a regular basis, similar to what `[/status, ...]` is doing for keeping in touch with the server.

This rare vs. standard scenario (in my eyes) would question indeed the need for OSC scheduling commands, and return to the question of whether not the standard bundle times can be considered as the audio-clock referencing times. Adding another scheduler on top of the existing scheduler certainly looks like a way more complex approach than simply providing a switch for the global server configuration once it's booted.

**Comment by telephon** [comment]
The standard size of timing information in OSC is 64 bit already.

http://opensoundcontrol.org/spec-1_0

> Time tags are represented by a 64 bit fixed point number. The first 32 bits specify the number of seconds since midnight on January 1, 1900, and the last 32 bits specify fractional parts of a second to a precision of about 200 picoseconds. This is the representation used by Internet NTP timestamps.

one could e.g. just switch the server mode to interpret the 64 bit as frame number. As I said, I don't see the for creating a clock on the server. That number repeats after around 6 million years, when you play at 96 KHz.

**Comment by sonoro1234** [comment]
I think we dont need anyting new, working at the same time with two clocks would be a nightmare.

In supernova we have two different clocks to choose: if you choose sampleclock your messages will be sample accurate (but will drift from anything synced to systemclok as some visuals in another program or any client using systemclock). If you choose systemclok you loose sample accuracy between messages but you get no drift and time accuracy with everything synced to systemclock.

There is no perfect solution, you have to work with one clock or the other. As I use osc messages for sequencing systemclock is the way to go.

scserver only has systemclock but would be easy to get also a sample clock to be choosen in some rare cases where you dont mind drifting from everything else but need sample accurate messages.

**Comment by telephon** [comment]
where is the documentation for this?  I couldn't find anything in the help system.

**Comment by jamshark70** [comment]
> where is the documentation for this?

If you mean supernova's clock option, I mentioned it days ago: "I think the problem is related to this: http://doc.sccode.org/Classes/ServerOptions.html#-useSystemClock (valid only for supernova btw)"

> I couldn't find anything in the help system.

I guess it isn't surprising that we wouldn't have extensive documentation on the lack of sample-accurate scheduling ;)

**Comment by gogobd** [comment]
Could anyone point me to the actual code of the sample clock in supernova?

**Comment by sonoro1234** [comment]
You can search for use_system_clock

**Comment by gogobd** [comment]
It's amazing how little code that actually is :-O

**Comment by muellmusik** [comment]
> I just wanted to note that syncing the language clock to the sample count would make these commands redundant.

I can see the case, albeit exotic, for sample accurate scheduling over a network. I would suggest a hybrid approach:

1. For the normal local case, a lang SampleClock syncs to the sample count.
2. For remote cases, we have a NetworkSampleClock which syncs using an OSC-based approach.
3. These could possibly be the same class and adjust depending on NetAddr, making code more reusable.

This would give the best possible performance in each case. Further considerations: LAN vs. WAN, single vs. multiple clients. For multiple over a LAN, a broadcast message might give best results, or perhaps a pseudo-reference beacon approach a la BeaconClock in Utopia. This avoids latency due to network contention.

Any approach to this needs to bear in mind network audio stuff such as AVB, and the possibility of sample count resets as well.

**Comment by Sciss** [comment]
> 2. For remote cases,

Please note that this is basically the case for any other client then sclang, unless we implement in the exact same fashion this clock in Python, Haskell, Scala, Clojure, ... which is basically impossible. So I consider case 2 to be very important and the most generic.

**Comment by LFSaw** [comment]
> unless we implement in the exact same fashion this clock in Python, Haskell, Scala, Clojure

wouldn't it be great to have that language-side code encapsulated in a primitive loaded as a library which can then be used in other clients as well?

**Comment by Sciss** [comment]
Yes, sure. But on the down-side it means platform neutral languages are suddenly not any longer write-once-write-everywhere (this goes at least for Python, Scala, Clojure, JavaScript, Ruby, probaby not Haskell which is compiled to native AFAIK).

**Comment by Sciss** [comment]
The big think of SC3 architecture is server-language separation, it comes with problems, but has led to many good things, especially the variety of clients. Introducing a new clock in a way that essentially undoes this separation is therefore not a good idea in my opinion, at least not if there is no API that works purely defined from the server side. So if we go back to Scott's list, I think there is nothing wrong with building a thing that can be shared by server and client, but it shouldn't be the only possibility to use this API.

**Comment by telephon** [comment]
the supernova implementation is server side, no? So we would need to negotiate that clock value between clients in a network, wouldn't we? I think I have lost the point, sorry, maybe someone can explain?

**Comment by Sciss** [comment]
I mean, as far as I understood the most recent arguments, the question is how the imagined protocol for providing a sample accurate clock is implemented. There are bascially two possibilities:

 1. the server is the "master" instance of maintaining a reference sample clock. Clients do not have to open their own audio interface connections. Instead, they keep their clock, which is still based on CPU clock, in sync by ticking the server and therefore maintaining an estimate of the sample clock by feedback of drift information.
 2. the functionality is implemented twice, once on the server and once on the client. In one respect, this could mean the process of keeping in sync is simpler, because we don't have a correction process for drift. In another respect, the pcoesses could become more complex, because now the client needs to essentially implement audio APIs.

Number 1 is needed if client and server have no means to poll the exact same sample clock. This is true for networked situations, but also, as I mentioned, if the client does not use the same C interface as the server, or if it does not _want_ to be infected by native C foreign interfaces (which is true for all languages that have their own runtimes and virtual machines). Another big question mark for number 2, as far as I can see, is that this potentially complicates the system for the user; for example, on Linux, it would mean that we now have two Jack clients, and I don't know Jack very well, but perhaps there is even no guarantee that both clients will at all times observe the same sample clock (e.g. if there is buffer (?) dropped).

This is why in my opinion number 1 should be the primary focus. It leaves the hardware interaction where it should be, with the server, and doesn't compromise the universal client-server separation. Also, without having ever built a prototype, it will be pure speculation as to what latencies are to expect, how well the servo mechanism works etc. I can imagine that with a proper algorithm latencies are not improved by number 2 but will still be dominated by OSC latency.

**Comment by Sciss** [comment]
@muellmusik can you explain how BeaconClock in Utopia works?

**Comment by telephon** [comment]
thanks, this is clearer now.

**Comment by muellmusik** [comment]
> So I consider case 2 to be very important and the most generic.

Note that with AVB, dante and other network audio growing rapidly, for LAN situations a shared interface (and thus samplecount within some window, etc) will become common.

> @muellmusik can you explain how BeaconClock in Utopia works?

The basic idea is that the single biggest issue in network latency – and more importantly, jitter in network latency – is network contention. The beacon approach seeks to avoid this through like so:

1) peers take turns sending a broadcast message with a 'beacon'.
2) the other peers then contact each other to compare what they thought the time was when they 'saw' the beacon, and then use some correction algorithm to adjust the differences and sync. (In BeaconClock the highest time wins, which simplifies certain things, but other approaches are possible)

Network contention is irrelevant since the beacon itself contains no timing info. Transport latency (which is usually minute on a LAN) is also irrelevant, since a broadcast message will arrive near simultaneously at all peers. (Well close enough for music). The approach is non-centralised, and does not require a time server. Further refinements include time-stamping the incoming message in the lowest possible layer. I had to fix arrival stamping in SC (it was actually happening when the message was processed by the lang, which is next to useless if there's lang blocking), and tried to figure out how to do this, but boost seems to keep you away from the lower network layers, unsurprisingly. But for musical sync it's close enough. For sample accurate sync, maybe not... (I should say I'm indebted to David Ogborn for teaching me about this, which I think has its origins in things like mine sensors.)

Anyway, this isn't directly applicable to what we're discussing, but might be adaptable in some way in LAN situations, e.g. using a broadcast message to send out sync. 

Of course what processes are we trying to sync?

If we just want the clock to stay roughly in time and serve up new server commands, then as long as messages arrive sufficiently ahead of time then a fair amount of jitter and latency are not a problem really.

If we're trying to sync other processes (visuals, or?) then it might be less tolerant. I think getting real 'sample accuracy' with that will be very hard, but again for most common purposes it would be close enough.

**Comment by jamshark70** [comment]
> I mean, as far as I understood the most recent arguments, the question is how the imagined protocol for providing a sample accurate clock is implemented. There are bascially two possibilities: 1. the server is the "master" instance of maintaining a reference sample clock. Clients do not have to open their own audio interface connections. Instead, they keep their clock, which is still based on CPU clock, in sync by ticking the server and therefore maintaining an estimate of the sample clock by feedback of drift information. ...

I agree. It's not reasonable to expect the client to connect to the audio device as well.

My concern with the `/sampleclock` idea remains: if the audio hardware is running slightly faster than the system clock, then the client will drift further and further behind, until *all* messages to the server are late, and at that point, you don't even have control-block accuracy, let alone sample accuracy. Adjusting for the drift seems necessary to me (hence my comment that `/sampleclock` seems a workaround). If non-SC clients want to support sample accuracy, they would have to implement mechanisms to correct the drift. (But they could also simply not implement it, and advertise that they are not sample accurate.)

Measuring empirically:

```
(
a = List.new;
o = OSCFunc({ |msg|
	if(a.size >= 200) {
		o.free;
		"done".postln;
	} {
		a.add(msg.last);
	}
}, '/status.reply', s.addr);
)

[a.minItem, a.maxItem, a.mean, a.median] - 44100
-> [ -0.049193728693353, 3.0830490475637, 0.49721934300032, 0.27003712338046 ]

[a.minItem, a.maxItem, a.mean, a.median] / 44100
-> [ 0.99999888449595, 1.0000699104092, 1.000011274815, 1.0000061232908 ]
```

If the maximum drift in either direction is 3 samples, the maximum correction would be 6.8027210884354e-05 seconds of correction for every second of real time. So it wouldn't be necessary to correct the drift very often -- even once every 10 seconds should be more than enough. So I don't think it would be burdensome for the server to send sample timing information back to the client, and the client does with it what it will.

**Comment by brianlheim** [comment]
just for bookkeeping—could we close #280 as a duplicate of this?

**Comment by sonoro1234** [comment]
summarizing: To provide this in scsynth a sample clock should be provided besides the system clock. (what happens in supernova now). Clients that want to use this sample clock without drift should implement some servo mechanism.
